File: #85_Maximal_Rectangle/Solution.java
import java.util.*;

public class Solution {
    public int maximalRectangle(char[][] matrix) {
        if (matrix == null || matrix.length == 0) return 0;
        int rows = matrix.length;
        int cols = matrix[0].length;
        int[] height = new int[cols];
        int maxArea = 0;
        for (int i = 0; i < rows; i++) {
        	for (int j = 0; j < cols; j++) {
        		if (matrix[i][j] == '1') height[j] += 1;
        		else height[j] = 0;
        	}
        	maxArea = Math.max(maxArea, findMaxArea(height));
        }
        return maxArea;
    }

    private int findMaxArea(int[] height) {
    	if (height == null || height.length == 0) return 0;
    	int[] h = Arrays.copyOf(height, height.length + 1);
    	int maxArea = 0;
    	Deque<Integer> stack = new LinkedList();
    	int k = 0;
    	while (k < h.length) {
    		if (stack.isEmpty() || h[stack.peek()] <= h[k]) {
    			stack.push(k++);
    		}
    		else {
    			int t = stack.pop();
    			int leftIndex = (stack.isEmpty()) ? -1 : stack.peek();
    			int rightIndex = k;
    			int tmpArea = h[t] * (rightIndex - leftIndex - 1);
    			maxArea = Math.max(maxArea, tmpArea);
    		}
    	}
    	return maxArea;
    }

    public static void main(String[] args) {
    	char[][] t = {
    		{'0', '0', '1', '1', '0', '1'},
    		{'0', '1', '0', '0', '0', '0'},
    		{'1', '0', '1', '0', '1', '0'}
    	};
    	Solution sol = new Solution();
    	System.out.println(sol.maximalRectangle(t));
    }
}

File: #85_Maximal_Rectangle/Solution.py
'''
O(m * n) Solution

for each row, find the maximum rectangle area
'''

class Solution:
    # @param {character[][]} matrix
    # @return {integer}
    def maximalRectangle(self, matrix):
        if not matrix or not matrix[0]:
            return 0
        height = [0 for j in range(len(matrix[0]))]
        ans = 0
        for i in range(len(matrix)):
            for j in range(len(matrix[0])):
                if matrix[i][j] == '0':
                    height[j] = 0
                else:
                    height[j] += 1
            ans = max(ans, self.largestRectangleArea(height))
        return ans

    # @param {integer[]} height
    # @return {integer}
    def largestRectangleArea(self, height):
        if not height:
        	return 0
        left = [0 for x in range(len(height))]
        right = [0 for x in range(len(height))]
        stk = []
        for x in range(len(height)):
        	while stk:
        		if height[x] > height[stk[len(stk) - 1]]: # watch for equal case
        			left[x] = stk[len(stk) - 1]
        			stk.append(x)
        			break
        		else:
        			stk.pop()
        	if not stk:
        		left[x] = -1
        		stk.append(x)
        stk = []
        for x in range(len(height) - 1, -1, -1):
        	while stk:
        		if height[x] > height[stk[len(stk) - 1]]: # watch for equal case
        			right[x] = stk[len(stk) - 1]
        			stk.append(x)
        			break
        		else:
        			stk.pop()
        	if not stk:
        		right[x] = len(height)  # right side should be len(height)
        		stk.append(x)
        ans = 0
        for x in range(len(height)):
        	ans = max(ans, height[x] * (right[x] - left[x] - 1))
        return ans

sol = Solution()
t = [['1', '0', '1'], ['1', '1', '1'], ['1', '1', '1']]
print sol.maximalRectangle(t)